app/services/duplicate_user_finder_service.rb

```diff
@@ -0,0 +1,23 @@
+class DuplicateUserFinderService
+  attr_reader :user
+
+  def initialize(user)
+    @user = user
+  end
+
+  def perform
+    same_phone = @user.phone_number.present? ? User.where(phone_number: @user.phone_number) : User.none
+    duplicate_data = if @user.birth_date.present? && @user.first_name.present? && @user.last_name.present?
+                       User.where(first_name: @user.first_name.capitalize, last_name: @user.last_name.upcase, birth_date: @user.birth_date)
+                     else
+                       User.none
+                     end
+
+    User.find_by(email: @user.email) ||
+      same_phone.or(duplicate_data)
+                .left_joins(:rdvs)
+                .group(:id)
+                .order('COUNT(rdvs.id) DESC')
+                .first
+  end
```

Comment:
The perform method is still quite hard to read imho. Here is a proposition for a rewrite. it feels simpler, but it doesn't do exactly the same as what you did. It sill passes the tests though. I'd be happy to discuss it live :)

```
  similar_users = User.where(email: @user.email)
    if @user.phone_number.present?
      similar_users = similar_users.or(User.where(phone_number: @user.phone_number))
    end
    if @user.birth_date.present? && @user.first_name.present? && @user.last_name.present?
      similar_users = similar_users.or(
        User.where(
          first_name: @user.first_name.capitalize,
          last_name: @user.last_name.upcase,
          birth_date: @user.birth_date
        )
      )
    end

    similar_users.left_joins(:rdvs).group(:id).order('COUNT(rdvs.id) DESC').first
```

Reply:
`email` is unique, what about a return clause ?

```
return User.find_by(email: @user.email) unless User.find_by(email: @user.email).nil?
```

Reply:
oh right if email is unique then it makes perfect sense to use a return clause ! 
just to avoid to do the query twice and be slightly DRYer you can do this in 2 lines: 

```
user_with_same_email = User.find_by(email: @user.email)
return user_with_same_email if user_with_same_email.present?
```